Introducing NRules.Integration.SimpleInjector

driekus77 opened the pull request:
We are using SimpleInjector as DI / IoC library and are starting to use NRules. 
For this I created this PR and Looked at, and copied from the Autofac integration.
Created an additional SimpleInjectorRuleRepositoryFactory class for named caching of RuleRepositories.

henryalliander committed:
[WIP] Made a start with NRules.Integration.SimpleInjector

driekus77 committed:
Work now ready for review.
Looked at, and copied from the Autofac integration.
Created an additional SimpleInjectorRuleRepositoryFactory class for named caching of RuleRepositories.

snikolayev commented:
@driekus77 I'm conflicted on this PR. I'm open to working through this with you, but first, I want to make sure it fits the intent of NRules integration packages. The idea is to supply integrations with other frameworks (e.g. IoC frameworks) that address the common use case for the majority of users of those frameworks. Looking at the proposed implementation, I'm not sure it serves that purpose.
I see you are implementing a named registry of rule repositories, using a container that considers named/keyed registrations an antipattern. I'm sure this is something you require for your application, but not sure this is how most users of SimpleInjector would do it. 
I imagine a vanilla case as something like this: 
* Find and register rules
* Register IRuleActivator and IDependencyResolver
* Register RuleRepository, set activator and load found rules into it
* Register ISessionFactory by resolving IRuleRepository, and compiling the rules stored in it; autowire IDependencyResolver
* Register ISession as transient, by resolving ISessionFactory and creating new session from it

IMO your proposed implementation is intended for a more specialized use case.
But let me know what you think.

driekus77 commented:
@snikolayev First of all thanks for your time. 

Secondly:
- I know that named/keyed registration is an anti-pattern (https://docs.simpleinjector.org/en/latest/howto.html#resolve-instances-by-key). But could get the "normal" way through TService as an interface NOT working.
- The way I managed to get it working was with inheriting from the Concrete RuleRepository. I liked this even less than named/keyed IoC resolving.

I will investigate some more to get it working like:
 => TService = IMyRules01 (interface which implements IRuleRepository interface) ==> TImplementation (Concrete) RuleRepository with  namespace prefix A.
 => TService = IMyRules02 (interface which implements IRuleRepository interface) ==> TImplementation (Concrete) RuleRepository with namespace prefix B.

Probably should wrap the RuleRepository and inherit this wrapper from TService....?

driekus77 committed:
Merge branch 'develop' into features/integration_simpleinjector

snikolayev commented:
@driekus77 I'm not particularly knowledgeable in SimpleInjector - I've never used it before, and I just read the docs while looking at your PR.

On your specific problem - may I ask why do you need named repositories? Why not load all rules into the same RuleRepository, just different named rule sets? Then you can compile different subset of rules into different ISessionFactory.

```c#
var repository = new RuleRepository();
repository.Load(x => x
    .From(...)
    .To("RuleSetName")
);

var compiler = new RuleCompiler();
var factory = compiler.Compile(repository.GetRuleSets().Where(rs => rs.Name == "RuleSetName"));
```

My main point here - I would find a way to use SimpleInjector in an idiomatic way, instead of fighting it. But you certainly know your use case much better.

On this PR - this is my main point - what you need here with multiple different sets of rules broken up by namespace, and how that drives the creation of different RuleRepositories, seems pretty unique to your use case, and I'm not sure how applicable it for a general purpose library published together with NRules. But let's see if we can coverge on a generic implementation that can address both.

driekus77 committed:
Removed Named/Keyed RuleRepositoryFactory in favor of using generic interface "Tag" type parameter for both IRuleRepository<...> and ISessionFactory<...>.

It uses Wrapper classes to deliver the correct impl/concrete type.

driekus77 commented:
Registration is no simpeler using a generic Interface type parameter as (sort of) "Tag" .
```csharp
            builder
                .RegisterRuleRepository<IMyRuleRepositoryA>(
                    x => x.Type(typeof(RuleWithConstructorDependency)))
                .RegisterRuleActivator()
                .RegisterDependencyResolver()
                .RegisterSessionFactory<IMyRuleRepositoryA>();
```

Resolving:
```csharp
 var sessionFactory = container.ResolveSessionFactory<IMyRuleRepositoryA>();
            var session = sessionFactory.CreateSession();
            session.Insert(new TestFact1());
            int fired = session.Fire();
```
Next to the generic versions there are also default Register/Resolve functions without a generic type parameters.

snikolayev commented:
@driekus77 the more I'm reading about Simple Injector, the more I'm getting convinced there is no generally reusable way to implement keyed/named registrations in SI that I would want to support as part of the NRules repo. As we discussed earlier, SI does not allow it, so you have to work around this limitation. My problem with these workarounds is that they are use case specific. I don't like the static factory approach, and, like you, I don't like this tag interface/wrapper approach even more. I particularly dislike the subclassing of RuleRepository and having to implement the decorator of ISessionFactory.
Don't get me wrong - I understand you have to use SI in this particular way, because that's what you use in your application/project and you need multiple different instances of the engine with different rules. I just don't want to have a specialized implementation as part of this repo.

driekus77 commented:
@snikolayev Again thanks for your time.

I think you are right and I'm fighting both NRules and SimpleInjector designs. Back to the drawing board. 

I am so used to DI/IoC preparing everything for me that I forget other ways to get the job done.

Basic use case for me would be registering, and preferable also compiling, one or more set of rules. This because we have a plugin/package system with SimpleInjector that has a strict separation of Registering and Resolving.
Runtime/Resolve-time we would need some way of addressing the ruleset. This could be an ISession if I follow NRules RuleSpec design correctly.  So perferable solution would be to let:
Register-time:
- Register all scanned rules
- Register IDependencyResolver and IRuleActivator
- Register IRuleRepository and ISessionFactory with optional RuleSet(s) specification
- Compile all Rules in Repository.

Runtime / Resolve-time:
- Inject ISessionFactory of even ISession and optinaly address a ruleset => How?

Currently I don't see yet how to address a (sub)set of rules through DI.... any idea's?
Maybe I'm to hard on RuleSet(s) vs Addressing through DI....

I will try to get back to this as soon as possible.

driekus77 committed:
Removed Interface tagging and added almost all the same Register functions as used with Autofac.

Next to that created an RuleSetsWithoutDependencyInjectionTests.cs to verify if I understand the "NRules" way of Loading, Compiling, and Firing Multiple RuleSets.

driekus77 commented:
@snikolayev Removed Interface tagging and added almost all the same Register functions as used with Autofac.

Next to that created an RuleSetsWithoutDependencyInjectionTests.cs to verify if I understand the "NRules" way of Loading, Compiling, and Firing Multiple RuleSets corectly?

Still thinking about addressing Multiple RuleSets through DI...

driekus77 committed:
Small fix in unittest

snikolayev commented on src/NRules.Integration/NRules.Integration.SimpleInjector/NRules.Integration.SimpleInjector.Tests/RuleSetsWithoutDependencyInjectionTests.cs:
```diff
@@ -0,0 +1,52 @@
+
+namespace NRules.Integration.SimpleInjector.Tests;
+
+public class RuleSetsWithoutDependencyInjectionTests
```
@driekus77 while this was helpful for you to understand how it works w/o DI, please delete before merging the PR as it's not relevant to this project

snikolayev replied:
And please remove fact types, rule types that it uses

snikolayev commented on src/NRules.Integration/NRules.Integration.SimpleInjector/NRules.Integration.SimpleInjector/RegistrationExtensions.cs:
```diff
@@ -0,0 +1,137 @@
+public static class RegistrationExtensions
+{
+    /// <summary>
+    /// RegisterRuleRepository the IRuleRepository.
```
Please add levant description in XML docs

snikolayev commented on src/NRules.Integration/NRules.Integration.SimpleInjector/NRules.Integration.SimpleInjector/RegistrationExtensions.cs:
```diff
@@ -0,0 +1,137 @@
+    public static Container RegisterRuleRepository(
+        this Container container,
+        Action<IRuleTypeScanner> scanAction,
+        Lifestyle? lifestyle = null)
```
Why nullable? Does it make more sense to make it non-nullable and default to singleton?

snikolayev replied:
Same in other places in this file

driekus77 replied:
Lifestyle.Singleton , Lifestyle.Transient, and Lifestyle.Scoped are all reference types and can only have null as default parameter. Nulllable is enabled so I need ? else I get a compile error.

snikolayev commented on src/NRules.Integration/NRules.Integration.SimpleInjector/NRules.Integration.SimpleInjector/RegistrationExtensions.cs:
```diff
@@ -0,0 +1,137 @@
+    {
+        lifestyle ??= Lifestyle.Singleton;
+        
+        // Scan for Rule(s)
```
Please remove comments that are not providing any additional information to what you can get by reading the code

snikolayev commented on src/NRules.Integration/NRules.Integration.SimpleInjector/NRules.Integration.SimpleInjector/RegistrationExtensions.cs:
```diff
@@ -0,0 +1,137 @@
+        // Register all rules found
+        foreach (var ruleType in ruleTypes)
+        {
+            if (container.GetRegistration(ruleType) == null)
```
Why  check if it's already registered? If it's already registered - the user of the code should just not call this again

snikolayev replied:
Same in other places in this file

snikolayev commented on src/NRules.Integration/NRules.Integration.SimpleInjector/NRules.Integration.SimpleInjector/RegistrationExtensions.cs:
```diff
@@ -0,0 +1,137 @@
+     /// <param name="container">The SimpleInjector container.</param>
+     /// <param name="lifestyle">The lifestyle scope. Default Singleton.</param>
+     /// <returns></returns>
+    public static Container RegisterSessionFactory(
```
Please fix indentation in all files - 4 spaces

snikolayev commented on src/NRules.Integration/NRules.Integration.SimpleInjector/NRules.Integration.SimpleInjector/RegistrationExtensions.cs:
```diff
@@ -0,0 +1,137 @@
+        container.Register<ISession>(() => container.GetInstance<ISessionFactory>().CreateSession(), lifestyle);
+        return container;
+    }
+
```
Please remove dangling line break

snikolayev commented on src/NRules.Integration/NRules.Integration.SimpleInjector/NRules.Integration.SimpleInjector/SimpleInjectorDependencyResolver.cs:
```diff
@@ -0,0 +1,25 @@
+
+namespace NRules.Integration.SimpleInjector
+{
+    public class SimpleInjectorDependencyResolver : IDependencyResolver
```
Missing XML docs

snikolayev commented on src/NRules.Integration/NRules.Integration.SimpleInjector/NRules.Integration.SimpleInjector/SimpleInjectorDependencyResolver.cs:
```diff
@@ -0,0 +1,25 @@
+            _container = container;
+        }
+
+        public object Resolve(IResolutionContext context, Type serviceType)
```
Missing XML docs. If same as interface, use:
`/// <inheritdoc cref="IDependencyResolver.Resolve"/>`

snikolayev commented on src/NRules.Integration/NRules.Integration.SimpleInjector/NRules.Integration.SimpleInjector/SimpleInjectorRuleActivator.cs:
```diff
@@ -0,0 +1,43 @@
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="container"></param>
```
Missing XML docs

snikolayev commented on src/NRules.Integration/NRules.Integration.SimpleInjector/NRules.Integration.SimpleInjector/SimpleInjectorRuleActivator.cs:
```diff
@@ -0,0 +1,43 @@
+            _container = container;
+        }
+
+        public IEnumerable<Rule> Activate(Type type)
```
Missing XML docs. Or use inheritdoc

snikolayev commented on src/NRules.Integration/NRules.Integration.SimpleInjector/NRules.Integration.SimpleInjector.Tests/RuleDependencyInjectionTests.cs:
```diff
@@ -0,0 +1,140 @@
+    {
+       [Fact]
+        public void RuleRepository_Resolved_Returns()
+        {
```
Please fix code indentation in this file. 4 spaces

snikolayev commented on src/NRules.Integration/NRules.Integration.SimpleInjector/NRules.Integration.SimpleInjector.Tests/RuleDependencyInjectionTests.cs:
```diff
@@ -0,0 +1,140 @@
+                session.Fire();
+                //Assert
+                Assert.Equal("It's done", service.Status);
+            }
```
Not a big deal, but please add line breaks before act and assert

snikolayev commented:
I think this is much closer to what I can include in the repo. I started a code review and added some comments. Thank you!

driekus77 committed:
[WIP] Draft proposal CompiledRuleSets

driekus77 commented:
@snikolayev Thanks for your review comments. I will address them tomorrow. 

I've managed to get DI working with, what I called (Please sugest other naming if needed ;-)), CompiledResultSets.
You can Specify/Define a CompiledResultSets by:
```csharp
public class MyCompiledRuleSets02 : CompiledRuleSets<MyCompiledRuleSets02>
{
    public override void DefineSpecification(RuleBulkLoadSpecBuilder builder)
    {
        builder
            .Specify("SubSet01",
                scan => scan.AssemblyOf<TestFact1>().Where(rm => rm.Namespace!.EndsWith("SubNamespace01")))
            .Specify("SubSet02",
                scan => scan.AssemblyOf<TestFact1>().Where(rm => rm.Namespace!.EndsWith("SubNamespace02")));
    }
}
```

Register:
```csharp
     container.RegisterCompiledRuleSets<MyCompiledRuleSets02>();
```

Resolve:
```csharp
            var instance = container.GetCompiledRuleSets<MyCompiledRuleSets02>();
            session02A = instance.GetOrCreate();
```

To leave existing implementation alone I added a RuleBulkLoadSpec which is needed to divide:
- Specifing / Defining the RuleTypes (Both for RuleRepository and DI)  ==> Register time DI
- Loading the RuleDefinitions using an RuleActivator ==> Resolve / Runtime DI

It is still in draft. Suggestions are welcome. If needed I will split up this PR into 2?

snikolayev commented:
@driekus77 I'm going to be out for two weeks starting this weekend, and I'm pretty tight on time, so I won't be able to go into the details of this new proposal, so I suggest you separate that out from this PR. If we move expeditiosly on addressing the PR comments, we can probably get this PR over the finish line in one day I have left.

driekus77 commented:
@snikolayev Oke I will split-up the work. 

But I don't see any review comments yet? Did you finish your review?

snikolayev commented:
> But I don't see any review comments yet? Did you finish your review?

I missed to hit submit review. You should see it now

driekus77 committed:
Fixed the review comments:
- Indentation checked
- XML Documentation chedked
- Type already registered checks removed
